macros/parserNumberWithUnits.pl

```diff
@@ -37,11 +37,46 @@ =head1 DESCRIPTION
 We now call on the Legacy version, which is used by
 num_cmp to handle numbers with units.
 
+New units can be added at run time by using the newUnit option
+
+       $a = NumberWithUnits("3 apples",{newUnit=>'apples'});
+
+A new unit can either be a string, in which case the string is added as a
+new unit with no relation to other units, or as a hashreference
+
+       $newUnit = {name => 'bear',
+                   conversion => {factor =>3, m=>1}};
+       $a = NumberWithUnits("3 bear", {newUnit=>$newUnit});
+
+You can also define your own conversion hash.  In the above example one bear
+is three meters.  (See Units.pm for examples). 
+
+Finally, the newUnit option can also be an array ref containing any number of
+new units to add.  A common reason for doing this would be to add the plural
+version of the unit as an equilvalent unit.  E.G.
+
+      $newUnits = ['apple',{name=>'apples',conversion=>{factor=>1,apple=>1}}];
+      $a = NumberWithUnits("3 apples",{newUnit=>$newUnits});
+
+In this case both 3 apple and 3 apples would be considered correct.  
+
+Note:  English pluralization is suprisingly hard, so WeBWorK will make no 
+attempt to display a grammerically correct result.  
+
 =cut
 
 loadMacros('MathObjects.pl');
 
 sub _parserNumberWithUnits_init {
+  # We make copies of these hashes here because these copies will be unique to  # the problem.  The hashes in Units are shared between problems.  We pass
+  # the hashes for these local copies to the NumberWithUnits package to use
+  # for all of its stuff.  
+  %fundamental_units = %Units::fundamental_units;
+  %known_units = %Units::known_units;
```

Comment:
> Its not always clear to me why some variables are local and some are global in PG

Yes, I understand.  I think much of that is left over from before PG used packages, and that was basically the only option.

>  If I add `my`'s should I also add accessors?

The actually variables used are the `my` variables in `Parcer::Legacy::ObjectWithUnits` that you added in the first file at the top of the page.  The ones here are only used to get copies of the ones from `Units::` to pass to `initializeUnits()` in the line below.  That is the routine that actually saves the values in the `my variables that are used for the rest of the problem.  So these definitely *should* be local variables, not global ones.

If you want to give access to the ones in `Parser::Legacy::ObjectWithUnits`, then that is a different question.  I'm not sure it is necessary.  The available units are the the ones in `Units::known_units` plus the ones defined in the problem, so there shouldn't be much question about that.  I suppose other macro packages could define some that you didn't know about.